float32[] errors
c_msgs[] centroids
int32[] number_op